src/qforte/utils/qbt_tapering.py

```diff
@@ -0,0 +1,318 @@
+"""
+
+import qforte as qf
+import numpy as np
+from itertools import product
+
+def find_Z2_symmetries(hamiltonian, qiskit_order = False):
+
+    """
+    This function computes the Z2 symmetries of the Hamiltonian, identifies the ID numbers of the qubits to
+    be tapered off, provides the corresponding sign structures, and constructs the necessary unitary operators.
+    It is based on the qubit tapering approach reported in https://arxiv.org/abs/1701.08213.
+
+    Arguments:
+
+    hamiltonian: QubitOperator
+        The qubit Hamiltonian whose Z2 symmetries we want to find.
+
+    qiskit_order: bool
+        If True, the qubits to be tapered off will have the smallest id numbers possible, as is done in Qiskit.
+        If False, the qubits to be tapered off will have the largest id numbers possible. This is the default,
+        since it leads to less operations once the qubits are removed.
+    """
+
+    # The algorithm consists of the following steps:
+    # 1) Convert qubit Hamiltonian to binary matrix, called the parity check matrix E
+    # 2) Find the basis of the kernel of the parity check matrix, ker{E}
+    # 3) Find maximal Abelian subgroup of the basis of ker{E}
+    # 4) Construct unitary matrices that will be used in the qubit_tapering algorithm
+    # 5) Compute the various signs corresponding to the eigenvalues of the sigma_x operators that will be removed
+
+    # 1) Convert qubit Hamiltonian to binary matrix
+
+    ## Obtain number of qubits and number of Pauli strings contained in qubit Hamiltonian
+    n_qubits = hamiltonian.num_qubits()
+    n_strings = len(hamiltonian.terms())
+
+    ## Construct parity check matrix
+    prt_chck_mtrx = np.zeros((n_strings, 2 * n_qubits), dtype=np.uint8)
+    for i, string in enumerate(hamiltonian.terms()):
+        for pauli in string[1].gates():
+            XYZ_gate = pauli.gate_id()
+            trgt = pauli.target()
+            if XYZ_gate == 'Z':
+                prt_chck_mtrx[i, n_qubits - trgt - 1] = 1
+            elif XYZ_gate == 'X':
+                prt_chck_mtrx[i, 2 * n_qubits - trgt - 1] = 1
+            else:
+                prt_chck_mtrx[i, n_qubits - trgt - 1] = 1
+                prt_chck_mtrx[i, 2 * n_qubits - trgt - 1] = 1
+
+    # 2) Find the basis of the kernel of the parity check matrix, ker{E}
+    # See https://en.wikipedia.org/wiki/Kernel_(linear_algebra)#Computation_by_Gaussian_elimination
+
+    ## Augment the (n_strings) × (2 * n_qubits) parity check matrix by the (2 * n_qubits) × (2 * n_qubits) identity matrix
+
+    prt_chck_mtrx_aug = np.concatenate((prt_chck_mtrx, np.identity(2 * n_qubits, dtype=np.uint8)), axis=0)
+
+    ## Find column echelon form (CEF) of augmented matix.
+
+    idx = 0
+    for row in range(n_strings):
+        ### Find a non-zero row
+        if np.any(prt_chck_mtrx_aug[row,idx:] == 1):
+            ### Find positions of non-zero elements
+            count = np.argwhere(prt_chck_mtrx_aug[row,idx:] == 1)
```

Comment:
Why not move the `argwhere` up? Instead of the any check, you can check if the result of `argwhere` is empty or not.

Similarly, please rename the variable. This is not a count.

Reply:
I understand that I can replace the `any` check by checking if the result of `argwhere` is empty or not. I am wondering if there is a computational benefit for doing so. Stylistically, I would prefer the `any` check. It seems more natural to me.

I named the variable `count` because `count.shape[0]` is the number of non-zero elements in a given row. I could rename `count` to `pos` since, technically, it holds the positions of the non-zero elements in a given row. What do you think?

Reply:
On further examination, `prt_chck_mtrx_aug[row,idx:] == 1` is just `prt_chck_mtrx_aug[row,idx:]`. New suggestion: Save that to a new variable and pass that to the two places you call `prt_chck_mtrx_aug[row,idx:] == 1`. My concern that motivated moving `argwhere` up was the duplication of that `prt_chck_matrix_aug` call. The new proposal also eliminates that.

While `pos` is technically correct, it's even less descriptive. Name it something like `nonzero_indices`.

Reply:
Good catch regarding `prt_chck_mtrx_aug[row,idx:] == 1`. Could you please explain the benefit of storing `prt_chck_mtrx_aug[row,idx:]` into a new variable instead of passing it twice as is? The reason I am asking is that if we decide to make this change, I will need to make similar modifications to many places in the code.

Reply:
I don't expect a significant performance impact; this is pure readability.

Reply:
In that case, do you mind if I just pass `prt_chck_mtrx_aug[row,idx:]` and similar for all other occurrences? I am afraid that creating new variables might be confusing when someone reads the code for the first time.

Reply:
As somebody who has been reading the code for the first time, the current approach is more confusing. I'm still not clear on what `prt_chck_mtrx_aug[row,idx:]` _means_.

Reply:
Thank you for your input.

I know that most of what I will right is already known to you, but for the sake of completeness I will still write it.

In that part of the code, I am finding the column echelon form of the augmented parity check matrix (`prt_chck_mtrx_aug`). This is done by looping over the rows (`row`) of the matrix. The variable `idx` holds the index of the next column to be brought into column echelon form. Essentially, this implementation splits the parity check matrix into two submatrices. The first one is already in column echelon form. The second one is not. The variable `idx` marks the first column of the submatrix that still needs to be brought to echelon form.

My problem is how to come up with meaningful variable names for all places that need to be modified.

Reply:
Are you telling me that's your problem or are you asking me to propose better variable names?

Reply:
Why not both :-)?

In any case, I am all for making the code more readable, but, as I already stated, I find the idea of adding more variables confusing, at least to me. If you have some suggestions, feel free to share them with me. We should keep in mind that the code is not supposed to be a manual, in the sense that I am not expecting people to read the code and _learn_ how qubit tapering _is_ performed. Apart from the original paper, which is also cited in the comments, we have the notes and presentation for it.

I hope that, based on my previous explanation, _prt_chck_mtrx_aug[row,idx:]_ is clear.

Reply:
I do not understand what could be confusing about creating a variable and then using that variable on each of the next two lines of code.

It looks like you need me to take a more active role in this review than I was expecting. I can do that, but I have a _lot_ of other things on my to-do list right now, so expect my response times to drop. Just push what you have to GitHub, and I'll get to it when I need a break from the rest of my list.

Reply:
Let me start by saying that `I already created` a new variable and further documented the code based on my explanation of `prt_chck_mtrx_aug[row,idx:]` posted above. Hopefully this new explanation will clear any confusion.

I apologize if my replies came across as if "I need you to take a more active role in this review". That was not and still is not my intention. The reason I was open to suggestions for an extra variable name is that the code _makes perfect sense to me_ the way it is. My goal is to make the code more readable to `you`, since you represent someone who is reading the code for the first time.

Nevertheless, let me state my case. In my mind, an additional variable is redundant and, thus, confusing to `me`, since there is already a variable with a quite descriptive name, `prt_chck_mtrx_aug`. In that instance of the code, I am focusing on a slice of a given `row`, hence `prt_chck_mtrx_aug[row,idx:]`. I understand if one finds the name `idx` cryptic and confusing. It would make more sense to me to change the name of the variable `idx` to something more descriptive and meaningful than to store `prt_chck_mtrx_aug[row,idx:]` in a new variable. An additional reason that I do not like creating an unnecessary variable is that someone that reads the code will be forced to search where the variable is defined. I do understand that this is not an issue here, because the variable will be defined just a couple of lines above its usage, but still I find it bad practice. In principle an extra variable will also add to the memory cost. Again, I do understand that in this particular example the additional memory requirement is negligible.

I would say let us put an end to this exchange of messages on github, because it is quite time consuming and we both need to focus on more important things than a single variable, although it is nice that we are both passionate about it. As I said in the beginning, I already introduced a new variable, so we should be good to go. I will be pushing a commit soon.